File: backend/services/GradCAMService.js
const tf = require('@tensorflow/tfjs-node');
const sharp = require('sharp');
const path = require('path');
const fs = require('fs');

class GradCAMService {
  constructor() {
    this.model = null;
    this.lastConvLayer = null;
  }

  /**
   * Încarcă modelul TensorFlow pentru Grad-CAM
   */
  async loadModel(modelPath) {
    try {
      console.log('Loading model for Grad-CAM:', modelPath);
      this.model = await tf.loadLayersModel(`file://${modelPath}`);
      
      // Găsește ultimul layer convolutional pentru Grad-CAM
      this.lastConvLayer = this.findLastConvLayer();
      console.log('Last conv layer found:', this.lastConvLayer?.name);
      
      return true;
    } catch (error) {
      console.error('Error loading model for Grad-CAM:', error);
      return false;
    }
  }

  /**
   * Găsește ultimul layer convolutional din model
   */
  findLastConvLayer() {
    const layers = this.model.layers;
    for (let i = layers.length - 1; i >= 0; i--) {
      const layer = layers[i];
      if (layer.getClassName() === 'Conv2D') {
        return layer;
      }
    }
    return null;
  }

  /**
   * Preprocesează imaginea pentru model
   */
  async preprocessImage(imagePath, targetSize = [224, 224]) {
    try {
      // Citește și redimensionează imaginea
      const imageBuffer = await sharp(imagePath)
        .resize(targetSize[0], targetSize[1])
        .removeAlpha()
        .raw()
        .toBuffer();

      // Convertește la tensor
      const tensor = tf.tensor3d(
        new Uint8Array(imageBuffer),
        [targetSize[0], targetSize[1], 3],
        'int32'
      );

      // Normalizează valorile pixelilor la [0, 1]
      const normalized = tensor.cast('float32').div(255.0);
      
      // Adaugă dimensiunea batch
      const batched = normalized.expandDims(0);
      
      tensor.dispose();
      normalized.dispose();
      
      return batched;
    } catch (error) {
      console.error('Error preprocessing image:', error);
      throw error;
    }
  }

  /**
   * Generează Grad-CAM heatmap
   */
  async generateGradCAM(imagePath, classIndex = 0) {
    try {
      if (!this.model || !this.lastConvLayer) {
        throw new Error('Model not loaded or last conv layer not found');
      }

      // Preprocesează imaginea
      const inputTensor = await this.preprocessImage(imagePath);
      
      // Creează un sub-model până la ultimul layer convolutional
      const convModel = tf.model({
        inputs: this.model.input,
        outputs: this.lastConvLayer.output
      });

      // Creează un model pentru predicție finală
      const classifierModel = tf.model({
        inputs: this.lastConvLayer.output,
        outputs: this.model.output
      });

      // Calculează gradienții folosind GradientTape
      const gradCAMResult = tf.tidy(() => {
        return tf.variableGrads(() => {
          // Forward pass prin conv layers
          const convOutput = convModel.predict(inputTensor);
          
          // Forward pass prin classifier
          const predictions = classifierModel.predict(convOutput);
          
          // Extrage scorul pentru clasa specificată
          const classScore = predictions.gather([classIndex], 1);
          
          return classScore;
        }, [inputTensor]);
      });

      // Calculează gradienții pentru ultimul layer conv
      const grads = gradCAMResult.grads[inputTensor];
      const convOutput = convModel.predict(inputTensor);

      // Calculează importanța fiecărui canal
      const channelWeights = tf.mean(grads, [0, 1, 2]);

      // Calculează heatmap-ul prin înmulțirea cu importanța
      const heatmap = tf.sum(
        tf.mul(convOutput.squeeze([0]), channelWeights),
        [2]
      );

      // Aplică ReLU pentru a păstra doar valorile pozitive
      const reluHeatmap = tf.relu(heatmap);

      // Normalizează heatmap-ul
      const normalizedHeatmap = this.normalizeHeatmap(reluHeatmap);

      // Cleanup
      inputTensor.dispose();
      convOutput.dispose();
      grads.dispose();
      heatmap.dispose();
      reluHeatmap.dispose();
      convModel.dispose();
      classifierModel.dispose();

      return normalizedHeatmap;
    } catch (error) {
      console.error('Error generating Grad-CAM:', error);
      throw error;
    }
  }

  /**
   * Normalizează heatmap-ul la [0, 1]
   */
  normalizeHeatmap(heatmap) {
    const min = tf.min(heatmap);
    const max = tf.max(heatmap);
    const range = tf.sub(max, min);
    
    const normalized = tf.div(tf.sub(heatmap, min), range);
    
    min.dispose();
    max.dispose();
    range.dispose();
    
    return normalized;
  }

  /**
   * Convertește tensor-ul heatmap la imagine
   */
  async tensorToImage(heatmapTensor, originalImagePath, outputPath) {
    try {
      // Obține dimensiunile imaginii originale
      const { width, height } = await sharp(originalImagePath).metadata();
      
      // Convertește tensor-ul la array
      const heatmapData = await heatmapTensor.data();
      const [heatmapHeight, heatmapWidth] = heatmapTensor.shape;

      // Creează o imagine colorată din heatmap
      const coloredHeatmap = this.applyColorMap(heatmapData, heatmapWidth, heatmapHeight);

      // Redimensionează heatmap-ul la dimensiunea imaginii originale
      const resizedHeatmap = await sharp(coloredHeatmap, {
        raw: {
          width: heatmapWidth,
          height: heatmapHeight,
          channels: 3
        }
      })
      .resize(width, height)
      .png()
      .toBuffer();

      // Încarcă imaginea originală
      const originalImage = await sharp(originalImagePath).toBuffer();

      // Combină imaginea originală cu heatmap-ul
      const result = await sharp(originalImage)
        .composite([{
          input: resizedHeatmap,
          blend: 'multiply',
          opacity: 0.6
        }])
        .jpeg({ quality: 90 })
        .toFile(outputPath);

      heatmapTensor.dispose();
      
      return {
        success: true,
        outputPath,
        size: result.size
      };
    } catch (error) {
      console.error('Error converting tensor to image:', error);
      throw error;
    }
  }

  /**
   * Aplică color map pe heatmap (jet colormap)
   */
  applyColorMap(heatmapData, width, height) {
    const rgbData = new Uint8Array(width * height * 3);
    
    for (let i = 0; i < heatmapData.length; i++) {
      const value = heatmapData[i];
      const rgb = this.jetColorMap(value);
      
      rgbData[i * 3] = rgb.r;
      rgbData[i * 3 + 1] = rgb.g;
      rgbData[i * 3 + 2] = rgb.b;
    }
    
    return rgbData;
  }

  /**
   * Jet color map implementation
   */
  jetColorMap(value) {
    const clampedValue = Math.max(0, Math.min(1, value));
    
    let r, g, b;
    
    if (clampedValue < 0.25) {
      r = 0;
      g = 4 * clampedValue;
      b = 1;
    } else if (clampedValue < 0.5) {
      r = 0;
      g = 1;
      b = 1 - 4 * (clampedValue - 0.25);
    } else if (clampedValue < 0.75) {
      r = 4 * (clampedValue - 0.5);
      g = 1;
      b = 0;
    } else {
      r = 1;
      g = 1 - 4 * (clampedValue - 0.75);
      b = 0;
    }
    
    return {
      r: Math.round(r * 255),
      g: Math.round(g * 255),
      b: Math.round(b * 255)
    };
  }

  /**
   * Generează heatmap complet - funcția principală
   */
  async generateHeatmapFromModel(imagePath, modelPath, outputDir, options = {}) {
    try {
      const {
        classIndex = 0,
        opacity = 0.6,
        colorIntensity = 1.0
      } = options;

      console.log('Starting Grad-CAM generation...');
      
      // Încarcă modelul
      const modelLoaded = await this.loadModel(modelPath);
      if (!modelLoaded) {
        throw new Error('Failed to load model');
      }

      // Generează Grad-CAM
      const heatmapTensor = await this.generateGradCAM(imagePath, classIndex);
      
      // Creează calea de output
      const timestamp = Date.now();
      const outputPath = path.join(outputDir, `gradcam_heatmap_${timestamp}.jpg`);
      
      // Convertește la imagine și salvează
      const result = await this.tensorToImage(heatmapTensor, imagePath, outputPath);
      
      console.log('Grad-CAM heatmap generated successfully:', outputPath);
      
      return {
        success: true,
        heatmapPath: outputPath,
        originalPath: imagePath,
        size: result.size,
        method: 'grad-cam'
      };
      
    } catch (error) {
      console.error('Error in Grad-CAM generation:', error);
      return {
        success: false,
        error: error.message,
        method: 'grad-cam'
      };
    }
  }
}

module.exports = new GradCAMService();

File: backend/services/gradcam_service.py
import numpy as np
import cv2
import tensorflow as tf
from tensorflow import keras
import sys
import json
import os
from PIL import Image
import matplotlib.pyplot as plt
import matplotlib.cm as cm

class GradCAMService:
    def __init__(self, model_path=None):
        """Inițializează serviciul Grad-CAM"""
        self.model = None
        self.last_conv_layer_name = None
        
        if model_path and os.path.exists(model_path):
            self.load_model(model_path)
        else:
            # Folosim un model pre-antrenat pentru demonstrație
            self.model = self.create_simple_model()
    
    def create_simple_model(self):
        """Creează un model simplu pentru demonstrație"""
        # Pentru demonstrație, folosim MobileNetV2 pre-antrenat
        base_model = keras.applications.MobileNetV2(
            input_shape=(224, 224, 3),
            include_top=False,
            weights='imagenet'
        )
        
        # Adăugăm layers pentru detecția deepfake
        model = keras.Sequential([
            base_model,
            keras.layers.GlobalAveragePooling2D(),
            keras.layers.Dense(128, activation='relu'),
            keras.layers.Dropout(0.2),
            keras.layers.Dense(1, activation='sigmoid')
        ])
        
        self.last_conv_layer_name = base_model.layers[-1].name
        return model
    
    def load_model(self, model_path):
        """Încarcă un model antrenat"""
        try:
            self.model = keras.models.load_model(model_path)
            # Găsește ultimul layer convolutional
            for layer in reversed(self.model.layers):
                if len(layer.output_shape) == 4:  # Conv layer
                    self.last_conv_layer_name = layer.name
                    break
        except Exception as e:
            print(f"Eroare la încărcarea modelului: {e}")
            self.model = self.create_simple_model()
    
    def preprocess_image(self, image_path, target_size=(224, 224)):
        """Preprocesează imaginea pentru model"""
        try:
            # Încarcă imaginea
            img = keras.preprocessing.image.load_img(image_path, target_size=target_size)
            img_array = keras.preprocessing.image.img_to_array(img)
            img_array = np.expand_dims(img_array, axis=0)
            img_array = keras.applications.mobilenet_v2.preprocess_input(img_array)
            return img_array, img
        except Exception as e:
            print(f"Eroare la preprocesarea imaginii: {e}")
            return None, None
    
    def make_gradcam_heatmap(self, img_array, pred_index=None, eps=1e-8):
        """Generează heatmap-ul Grad-CAM"""
        if self.model is None or self.last_conv_layer_name is None:
            raise ValueError("Model sau layer conv nu sunt inițializate")
        
        try:
            # Creează un model care mapează input-ul la activările conv layer-ului și la predicții
            grad_model = keras.models.Model(
                inputs=[self.model.inputs],
                outputs=[self.model.get_layer(self.last_conv_layer_name).output, self.model.output]
            )
            
            # Calculează gradientele
            with tf.GradientTape() as tape:
                last_conv_layer_output, preds = grad_model(img_array)
                if pred_index is None:
                    pred_index = tf.argmax(preds[0])
                class_channel = preds[:, pred_index]
            
            # Gradientele feature map-ului ultimului conv layer cu privire la clasa predicată
            grads = tape.gradient(class_channel, last_conv_layer_output)
            
            # Vector în care fiecare intrare este importanța medie a gradientului pe un anumit feature map
            pooled_grads = tf.reduce_mean(grads, axis=(0, 1, 2))
            
            # Înmulțește fiecare canal din feature map cu importanța lui
            last_conv_layer_output = last_conv_layer_output[0]
            heatmap = last_conv_layer_output @ pooled_grads[..., tf.newaxis]
            heatmap = tf.squeeze(heatmap)
            
            # Normalizează heatmap-ul între 0 și 1
            heatmap = tf.maximum(heatmap, 0) / tf.math.reduce_max(heatmap)
            return heatmap.numpy()
            
        except Exception as e:
            print(f"Eroare la generarea Grad-CAM: {e}")
            # Returnează un heatmap sintetic pentru debugging
            return self.generate_synthetic_heatmap()
    
    def generate_synthetic_heatmap(self, size=(7, 7)):
        """Generează un heatmap sintetic pentru demonstrație"""
        # Creează zone de activare în diferite regiuni
        heatmap = np.zeros(size)
        
        # Zonă de activare în centru (față)
        center_x, center_y = size[0] // 2, size[1] // 2
        heatmap[center_x-1:center_x+2, center_y-1:center_y+2] = 0.8
        
        # Zone de activare în colțuri (ochi, gură)
        heatmap[1, 1] = 0.6  # ochi stâng
        heatmap[1, -2] = 0.6  # ochi drept
        heatmap[-2, center_y] = 0.5  # zona gurii
        
        return heatmap
    
    def save_gradcam_heatmap(self, img_path, heatmap, output_path, alpha=0.6):
        """Salvează heatmap-ul suprapus pe imaginea originală"""
        try:
            # Încarcă imaginea originală
            img = keras.preprocessing.image.load_img(img_path)
            img = keras.preprocessing.image.img_to_array(img)
            
            # Redimensionează heatmap-ul la dimensiunea imaginii
            heatmap_resized = cv2.resize(heatmap, (img.shape[1], img.shape[0]))
            heatmap_resized = np.uint8(255 * heatmap_resized)
            
            # Aplică colormap
            heatmap_colored = cm.jet(heatmap_resized)
            heatmap_colored = np.uint8(255 * heatmap_colored)
            
            # Convertește la RGB
            heatmap_rgb = cv2.cvtColor(heatmap_colored, cv2.COLOR_RGBA2RGB)
            
            # Suprapune heatmap-ul pe imaginea originală
            superimposed_img = heatmap_rgb * alpha + img * (1 - alpha)
            superimposed_img = np.uint8(superimposed_img)
            
            # Salvează rezultatul
            result_img = Image.fromarray(superimposed_img)
            result_img.save(output_path, 'JPEG', quality=95)
            
            return True
            
        except Exception as e:
            print(f"Eroare la salvarea heatmap-ului: {e}")
            return False
    
    def generate_heatmap(self, image_path, output_path, confidence_score=None):
        """Funcția principală pentru generarea heatmap-ului"""
        try:
            # Preprocesează imaginea
            img_array, original_img = self.preprocess_image(image_path)
            if img_array is None:
                return False
            
            # Generează predicția
            prediction = self.model.predict(img_array, verbose=0)
            predicted_class = int(prediction[0][0] > 0.5)
            confidence = float(prediction[0][0])
            
            # Generează heatmap-ul Grad-CAM
            heatmap = self.make_gradcam_heatmap(img_array, predicted_class)
            
            # Salvează heatmap-ul suprapus
            success = self.save_gradcam_heatmap(image_path, heatmap, output_path)
            
            return {
                'success': success,
                'confidence': confidence,
                'predicted_class': predicted_class,
                'heatmap_path': output_path if success else None
            }
            
        except Exception as e:
            print(f"Eroare la generarea heatmap-ului: {e}")
            return {'success': False, 'error': str(e)}

def main():
    """Funcția principală care poate fi apelată din Node.js"""
    if len(sys.argv) < 3:
        print(json.dumps({'error': 'Parametri insuficienți'}))
        return
    
    image_path = sys.argv[1]
    output_path = sys.argv[2]
    confidence_score = float(sys.argv[3]) if len(sys.argv) > 3 else None
    
    # Inițializează serviciul
    gradcam = GradCAMService()
    
    # Generează heatmap-ul
    result = gradcam.generate_heatmap(image_path, output_path, confidence_score)
    
    # Returnează rezultatul ca JSON
    print(json.dumps(result))

if __name__ == "__main__":
    main()
